<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Home · MaxEntNucleotideBiases.jl</title><script data-outdated-warner src="assets/warner.js"></script><link rel="canonical" href="https://adigioacchino.github.io/MaxEntNucleotideBiases.jl/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href>MaxEntNucleotideBiases.jl</a></span></div><form class="docs-search" action="search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>Home</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Home</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Home</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/adigioacchino/MaxEntNucleotideBiases.jl/blob/main/docs/src/index.md#" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="MaxEntNucleotideBiases"><a class="docs-heading-anchor" href="#MaxEntNucleotideBiases">MaxEntNucleotideBiases</a><a id="MaxEntNucleotideBiases-1"></a><a class="docs-heading-anchor-permalink" href="#MaxEntNucleotideBiases" title="Permalink"></a></h1><p>Documentation for <a href="https://github.com/adigioacchino/MaxEntNucleotideBiases.jl">MaxEntNucleotideBiases</a>.</p><ul><li><a href="#MaxEntNucleotideBiases.ComputeEntropy-Tuple{Int64, MaxEntNucleotideBiases.NucleotideModel}"><code>MaxEntNucleotideBiases.ComputeEntropy</code></a></li><li><a href="#MaxEntNucleotideBiases.ComputeLoglikelihood-Tuple{String, MaxEntNucleotideBiases.NucleotideModel}"><code>MaxEntNucleotideBiases.ComputeLoglikelihood</code></a></li><li><a href="#MaxEntNucleotideBiases.ComputeMinusEnergy-Tuple{AbstractString, MaxEntNucleotideBiases.NucleotideModel}"><code>MaxEntNucleotideBiases.ComputeMinusEnergy</code></a></li><li><a href="#MaxEntNucleotideBiases.ComputePressure-Tuple{Int64, MaxEntNucleotideBiases.NucleotideModel}"><code>MaxEntNucleotideBiases.ComputePressure</code></a></li><li><a href="#MaxEntNucleotideBiases.ComputeSymmetrizedKL-Tuple{Int64, MaxEntNucleotideBiases.NucleotideModel, MaxEntNucleotideBiases.NucleotideModel}"><code>MaxEntNucleotideBiases.ComputeSymmetrizedKL</code></a></li><li><a href="#MaxEntNucleotideBiases.EvalLogZ-Tuple{MaxEntNucleotideBiases.NucleotideModel, Int64}"><code>MaxEntNucleotideBiases.EvalLogZ</code></a></li><li><a href="#MaxEntNucleotideBiases.EvalLogZFast-Tuple{MaxEntNucleotideBiases.NucleotideModel, Int64}"><code>MaxEntNucleotideBiases.EvalLogZFast</code></a></li><li><a href="#MaxEntNucleotideBiases.ForcesDict-Tuple{MaxEntNucleotideBiases.NucleotideModel}"><code>MaxEntNucleotideBiases.ForcesDict</code></a></li><li><a href="#MaxEntNucleotideBiases.GaugeMaskVariables-Tuple{Vector{String}}"><code>MaxEntNucleotideBiases.GaugeMaskVariables</code></a></li><li><a href="#MaxEntNucleotideBiases.GenerateTM-Tuple{MaxEntNucleotideBiases.NucleotideModel}"><code>MaxEntNucleotideBiases.GenerateTM</code></a></li><li><a href="#MaxEntNucleotideBiases.GenerateTMLast-Tuple{MaxEntNucleotideBiases.NucleotideModel}"><code>MaxEntNucleotideBiases.GenerateTMLast</code></a></li><li><a href="#MaxEntNucleotideBiases.MetropolisSampling-Tuple{Int64, MaxEntNucleotideBiases.NucleotideModel}"><code>MaxEntNucleotideBiases.MetropolisSampling</code></a></li><li><a href="#MaxEntNucleotideBiases.ModelFit"><code>MaxEntNucleotideBiases.ModelFit</code></a></li><li><a href="#MaxEntNucleotideBiases.ZerosumGauge-Tuple{MaxEntNucleotideBiases.NucleotideModel}"><code>MaxEntNucleotideBiases.ZerosumGauge</code></a></li><li><a href="#MaxEntNucleotideBiases._ComputeNObs-Tuple{Vector{String}, Vector{String}, Int64}"><code>MaxEntNucleotideBiases._ComputeNObs</code></a></li><li><a href="#MaxEntNucleotideBiases._MCStep!-Tuple{Int64, Int64, Vector{String}, MaxEntNucleotideBiases.NucleotideModel, Float64}"><code>MaxEntNucleotideBiases._MCStep!</code></a></li><li><a href="#MaxEntNucleotideBiases._PreprocessSeqs-Tuple{Vector{String}}"><code>MaxEntNucleotideBiases._PreprocessSeqs</code></a></li></ul><article class="docstring"><header><a class="docstring-binding" id="MaxEntNucleotideBiases.ComputeEntropy-Tuple{Int64, MaxEntNucleotideBiases.NucleotideModel}" href="#MaxEntNucleotideBiases.ComputeEntropy-Tuple{Int64, MaxEntNucleotideBiases.NucleotideModel}"><code>MaxEntNucleotideBiases.ComputeEntropy</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">ComputeEntropy(L::Int, model::NucleotideModel; fast::Bool=false)</code></pre><p>Compute the entropy of the model with parameters <code>model_pars</code> and having length <code>L</code>. The computation exploit the derivatives of the partition function, that can be approximated to make it faster (using <code>fast=true</code>), although it should be very quick in any case.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/adigioacchino/MaxEntNucleotideBiases.jl/blob/1d5f1578c084fbda69c4bd10d5d549c8beae6170/src/MENB_Utils.jl#L120-L127">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MaxEntNucleotideBiases.ComputeLoglikelihood-Tuple{String, MaxEntNucleotideBiases.NucleotideModel}" href="#MaxEntNucleotideBiases.ComputeLoglikelihood-Tuple{String, MaxEntNucleotideBiases.NucleotideModel}"><code>MaxEntNucleotideBiases.ComputeLoglikelihood</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">ComputeLoglikelihood(seq::String, model::NucleotideModel; logZ=missing)</code></pre><p>Given a sequence <code>seq</code> and the model parameters <code>model_pars</code>, compute the log-likelihood (energy minus log of Z) of this sequence. logZ can be passed directly if pre-computed, otherwise is it computed each time this function is called.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/adigioacchino/MaxEntNucleotideBiases.jl/blob/1d5f1578c084fbda69c4bd10d5d549c8beae6170/src/MENB_Utils.jl#L28-L34">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MaxEntNucleotideBiases.ComputeMinusEnergy-Tuple{AbstractString, MaxEntNucleotideBiases.NucleotideModel}" href="#MaxEntNucleotideBiases.ComputeMinusEnergy-Tuple{AbstractString, MaxEntNucleotideBiases.NucleotideModel}"><code>MaxEntNucleotideBiases.ComputeMinusEnergy</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">ComputeMinusEnergy(seq::String, model::NucleotideModel)</code></pre><p>Given a sequence <code>seq</code> and the model parameters <code>model_pars</code>,  compute (minus) the energy of this sequence.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/adigioacchino/MaxEntNucleotideBiases.jl/blob/1d5f1578c084fbda69c4bd10d5d549c8beae6170/src/MENB_Utils.jl#L1-L6">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MaxEntNucleotideBiases.ComputePressure-Tuple{Int64, MaxEntNucleotideBiases.NucleotideModel}" href="#MaxEntNucleotideBiases.ComputePressure-Tuple{Int64, MaxEntNucleotideBiases.NucleotideModel}"><code>MaxEntNucleotideBiases.ComputePressure</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">ComputePressure(L::Int, model::NucleotideModel; fast::Bool=false)</code></pre><p>The pressure on a genome can be quantified as how the genome is different from a random uniform one. Therefore, this function compute the difference in the entropy of a random uniform model and the model described with <code>model_pars</code>, with lenght <code>L</code>, rescaled with the entropy of the uniform model. Setting <code>fast=true</code> allows for a quicker estimation of the entropy of the model (it can be useful for extremely large sequences).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/adigioacchino/MaxEntNucleotideBiases.jl/blob/1d5f1578c084fbda69c4bd10d5d549c8beae6170/src/MENB_Utils.jl#L146-L154">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MaxEntNucleotideBiases.ComputeSymmetrizedKL-Tuple{Int64, MaxEntNucleotideBiases.NucleotideModel, MaxEntNucleotideBiases.NucleotideModel}" href="#MaxEntNucleotideBiases.ComputeSymmetrizedKL-Tuple{Int64, MaxEntNucleotideBiases.NucleotideModel, MaxEntNucleotideBiases.NucleotideModel}"><code>MaxEntNucleotideBiases.ComputeSymmetrizedKL</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">ComputeSymmetrizedKL(L::Int, model1::NucleotideModel, 
                    model2::NucleotideModel; fast::Bool=false)</code></pre><p>Compute the symmetrized version of the Kullback-Leibler divergence between the distributions defined by <code>model1_pars</code> and <code>model2_pars</code>. Setting <code>fast=true</code> allows for a quicker estimation  of the KL divergence (it can be useful for extremely large sequences).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/adigioacchino/MaxEntNucleotideBiases.jl/blob/1d5f1578c084fbda69c4bd10d5d549c8beae6170/src/MENB_Utils.jl#L160-L167">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MaxEntNucleotideBiases.EvalLogZ-Tuple{MaxEntNucleotideBiases.NucleotideModel, Int64}" href="#MaxEntNucleotideBiases.EvalLogZ-Tuple{MaxEntNucleotideBiases.NucleotideModel, Int64}"><code>MaxEntNucleotideBiases.EvalLogZ</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">EvalLogZ(model::NucleotideModel, L::Int)</code></pre><p>Compute the partition function of a model of length L through the transfer matrix method.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/adigioacchino/MaxEntNucleotideBiases.jl/blob/1d5f1578c084fbda69c4bd10d5d549c8beae6170/src/MENB_PartitionFunction.jl#L86-L91">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MaxEntNucleotideBiases.EvalLogZFast-Tuple{MaxEntNucleotideBiases.NucleotideModel, Int64}" href="#MaxEntNucleotideBiases.EvalLogZFast-Tuple{MaxEntNucleotideBiases.NucleotideModel, Int64}"><code>MaxEntNucleotideBiases.EvalLogZFast</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">EvalLogZFast(model::NucleotideModel, L::Int)</code></pre><p>Compute the partition function of a model of length L by taking the real part of the largest eigenvalue of the transfer matrix.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/adigioacchino/MaxEntNucleotideBiases.jl/blob/1d5f1578c084fbda69c4bd10d5d549c8beae6170/src/MENB_PartitionFunction.jl#L111-L116">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MaxEntNucleotideBiases.ForcesDict-Tuple{MaxEntNucleotideBiases.NucleotideModel}" href="#MaxEntNucleotideBiases.ForcesDict-Tuple{MaxEntNucleotideBiases.NucleotideModel}"><code>MaxEntNucleotideBiases.ForcesDict</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">ForcesDict(model::NucleotideModel)</code></pre><p>Return a dictionary of the form motif =&gt; force.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/adigioacchino/MaxEntNucleotideBiases.jl/blob/1d5f1578c084fbda69c4bd10d5d549c8beae6170/src/MaxEntNucleotideBiases.jl#L23-L27">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MaxEntNucleotideBiases.GaugeMaskVariables-Tuple{Vector{String}}" href="#MaxEntNucleotideBiases.GaugeMaskVariables-Tuple{Vector{String}}"><code>MaxEntNucleotideBiases.GaugeMaskVariables</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">GaugeMaskVariables(motifs::Vector{String})</code></pre><p>This function returns a Array{Bool} corresponding to motifs that have to be inferred after all motifs containing Ts are fixed to 0 thanks to gauge transformations.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/adigioacchino/MaxEntNucleotideBiases.jl/blob/1d5f1578c084fbda69c4bd10d5d549c8beae6170/src/MENB_Gauge.jl#L4-L9">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MaxEntNucleotideBiases.GenerateTM-Tuple{MaxEntNucleotideBiases.NucleotideModel}" href="#MaxEntNucleotideBiases.GenerateTM-Tuple{MaxEntNucleotideBiases.NucleotideModel}"><code>MaxEntNucleotideBiases.GenerateTM</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">GenerateTM(model::NucleotideModel)</code></pre><p>Return the trasfer matrix used for the computation of the partition function. The first index correspond to the left-most nucleotide, the last index to the right-most one.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/adigioacchino/MaxEntNucleotideBiases.jl/blob/1d5f1578c084fbda69c4bd10d5d549c8beae6170/src/MENB_PartitionFunction.jl#L1-L7">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MaxEntNucleotideBiases.GenerateTMLast-Tuple{MaxEntNucleotideBiases.NucleotideModel}" href="#MaxEntNucleotideBiases.GenerateTMLast-Tuple{MaxEntNucleotideBiases.NucleotideModel}"><code>MaxEntNucleotideBiases.GenerateTMLast</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">GenerateTMLast(model::NucleotideModel)</code></pre><p>Return the last transfer matrix used for the computation of the partition function. The first index correspond to the left-most nucleotide, the last index to the right-most one.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/adigioacchino/MaxEntNucleotideBiases.jl/blob/1d5f1578c084fbda69c4bd10d5d549c8beae6170/src/MENB_PartitionFunction.jl#L41-L47">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MaxEntNucleotideBiases.MetropolisSampling-Tuple{Int64, MaxEntNucleotideBiases.NucleotideModel}" href="#MaxEntNucleotideBiases.MetropolisSampling-Tuple{Int64, MaxEntNucleotideBiases.NucleotideModel}"><code>MaxEntNucleotideBiases.MetropolisSampling</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">MetropolisSampling(L::Int, model::NucleotideModel; 
                         beta::Float64=1.0, Nsamples::Int=1, Nsteps::Int=1, Ntherm::Int=L*10, 
                         startseq::Union{String,Missing}=missing)</code></pre><p>Sample Nsample sequences of length L at inverse temperature beta using a model given in model.  Ntherm is the number of MonteCarlo steps done before starting collecting sequences,  startseq is the starting sequence.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/adigioacchino/MaxEntNucleotideBiases.jl/blob/1d5f1578c084fbda69c4bd10d5d549c8beae6170/src/MENB_Utils.jl#L82-L90">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MaxEntNucleotideBiases.ModelFit" href="#MaxEntNucleotideBiases.ModelFit"><code>MaxEntNucleotideBiases.ModelFit</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">ModelFit(seqs::Union{Vector{String}, String}, Lmotifs::Int, Lmodel::Union{Int,Missing}=missing; 
              add_pseudocount::Bool=false, tolerance::Float64=0.01, max_iter::Int=100, 
              verbose::Bool=false, fast::Union{Bool,String}=&quot;auto&quot;, ZS_gauge::Bool=true)</code></pre><p>Fit the model parameters, which are:</p><ul><li>only 1-point functions (fields) if Lmotifs==1;</li><li>1-point and 2-point functions (fields, 2-mer forces) if Lmotifs==2;</li><li>1-point, 2-point and 3-point functions (fields, 2-mer and 3-mer forces) if Lmotifs==3.</li></ul><p><code>Lmodel</code> is the number of nucleotides used for the inference. If <code>seqs</code> is made of  sequences of constant length (or if it is String and not a vector), the default option will use the sequences length as <code>Lmodel</code>; otherwise, the default behaviour is to take <code>Lmodel</code> = 5000.</p><p>If <code>add_pseudocount</code>, a single pseudocount is added for each observed number of nucleotides and dinucleotides.</p><p>If <code>fast</code>, the partition function is estimated through the top eigenvalue of the transfer matrix alone (much faster, but slightly less precise, expecially for short sequences). The default value, &quot;auto&quot;, automatically uses the fast  evaluation for long sequences.</p><p><code>tolerance</code> and <code>max_iter</code> are parameters for the Newton-Raphson algorithm  used to solve the system of equations.</p><p><code>ZS_gauge</code> specifies whether the result has to be put in the zero sum gauge before being returned.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/adigioacchino/MaxEntNucleotideBiases.jl/blob/1d5f1578c084fbda69c4bd10d5d549c8beae6170/src/MENB_InferModel.jl#L33-L61">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MaxEntNucleotideBiases.ZerosumGauge-Tuple{MaxEntNucleotideBiases.NucleotideModel}" href="#MaxEntNucleotideBiases.ZerosumGauge-Tuple{MaxEntNucleotideBiases.NucleotideModel}"><code>MaxEntNucleotideBiases.ZerosumGauge</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">ZerosumGauge(model_pars::NucleotideModel)</code></pre><p>This function takes as input a dictionary describing the parmeters of a model and changes the gauge into the &quot;zero-sum&quot; gauge.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/adigioacchino/MaxEntNucleotideBiases.jl/blob/1d5f1578c084fbda69c4bd10d5d549c8beae6170/src/MENB_Gauge.jl#L15-L20">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MaxEntNucleotideBiases._ComputeNObs-Tuple{Vector{String}, Vector{String}, Int64}" href="#MaxEntNucleotideBiases._ComputeNObs-Tuple{Vector{String}, Vector{String}, Int64}"><code>MaxEntNucleotideBiases._ComputeNObs</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">_ComputeNObs(seqs::Vector{String}, independent_motifs::Vector{String}, L::Int)</code></pre><p>For each motif in <code>independent_motifs</code>, compute the number of observed motifs in each sequence in <code>seqs</code>, then divide by the sequence length, take the average of these intensive fractions over the sequences, and multiply by the model length L.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/adigioacchino/MaxEntNucleotideBiases.jl/blob/1d5f1578c084fbda69c4bd10d5d549c8beae6170/src/MENB_InferModel.jl#L20-L26">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MaxEntNucleotideBiases._MCStep!-Tuple{Int64, Int64, Vector{String}, MaxEntNucleotideBiases.NucleotideModel, Float64}" href="#MaxEntNucleotideBiases._MCStep!-Tuple{Int64, Int64, Vector{String}, MaxEntNucleotideBiases.NucleotideModel, Float64}"><code>MaxEntNucleotideBiases._MCStep!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">_MCStep!(L::Int, mlk::Int, curr_seq::Vector{String}, 
              model::NucleotideModel, 
              beta::Float64)</code></pre><p>Performs a MonteCarlo step; L is curr<em>seq length, mlk is the maximum length of the motifs considered, curr</em>seq is the starting sequence, model specifies the model,  and beta is the inverse temperature.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/adigioacchino/MaxEntNucleotideBiases.jl/blob/1d5f1578c084fbda69c4bd10d5d549c8beae6170/src/MENB_Utils.jl#L51-L59">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="MaxEntNucleotideBiases._PreprocessSeqs-Tuple{Vector{String}}" href="#MaxEntNucleotideBiases._PreprocessSeqs-Tuple{Vector{String}}"><code>MaxEntNucleotideBiases._PreprocessSeqs</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">_PreprocessSeqs(seqs_in::Vector{String})</code></pre><p>Pre-process the vector of sequences <code>seqs_in</code>, so that sequences only contains letters A, C, G, T. The pre-processed sequence vector is the returned.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/adigioacchino/MaxEntNucleotideBiases.jl/blob/1d5f1578c084fbda69c4bd10d5d549c8beae6170/src/MENB_InferModel.jl#L1-L6">source</a></section></article></article><nav class="docs-footer"><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.24 on <span class="colophon-date" title="Saturday 11 March 2023 21:08">Saturday 11 March 2023</span>. Using Julia version 1.8.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
